test/parallel/test-http-socket-encoding-error.js

```diff
@@ -0,0 +1,23 @@
+'use strict';
+
+const common = require('../common');
+const http = require('http');
+
+const server = http.createServer().listen(0, connectToServer);
+
+server.on('connection', (socket) => {
+  common.expectsError(() => socket.setEncoding(''),
```

Comment:
I think you might still be testing the old `setEncoding` here. I think you might be better off just using normal request callback and then using `req.socket.setEncoding`.

Reply:
That `req` part isn't an option in the `'connection'` event. Are you saying to test with an incoming request _after_ the `'connection'`? This wouldn't throw an error because the request is parsed by then.

For example:
```
http.createServer( ( req, res ) => {
    req.socket.setEncoding( 'utf8' );
    res.end( 'OK' );
} )
```

is ok to do. It's only "pre-parse" attempts to manipulate the incoming buffer

Reply:
Yeah, you can ignore me. It's complaining about the fact that you're providing an argument to the error, as mentioned above by @BridgeAR.

But if you're saying it's ok then this PR isn't quite valid because calling that `req.socket.setEncoding('utf8')` will still throw as things exist currently.

Reply:
But also this isn't strictly correct. It's definitely not OK to call `req.socket.setEncoding` until after the `end` event.

Reply:
Hmm. Do you know any history behind `setEncoding` and how it even came about?

Right now I override `setEncoding` in the `connectionListenerInternal` method. This causes a fatal error as detailed in #18118. Are you recommending we remove it completely from the HTTP module?

Reply:
[This comment](https://github.com/nodejs/node/pull/18178#issuecomment-358728144) from @addaleax on the original PR suggests there are other uses of `setEncoding` when _possibly_ responding to a protocol change request or connection upgrade requests.

There's a slight chance it could be necessary when using abstract protocols such as `Internet Printing Protocol/1.0` or some other media protocols. Although, as I discussed in that PR, this should not be allowed in the scope of HTTP. HTTP in, HTTP out.

So this would require moving `socket.setEncoding = socketSetEncoding;` to somewhere in the `httpSocketSetup` method or something?

Reply:
@iSkore Thanks for the link. I might need to think about this more. I think that particular use-case would be addressed by resetting the socket `setEncoding` back within the the `upgrade` code path but there might be others that are not addressed. This might also affect how modules like `ws` operate. Let me do some research or feel free to do it yourself — but it seems like this PR will need to be updated to account for at least `upgrade`, if not more.